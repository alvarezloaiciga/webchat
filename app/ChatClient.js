// @flow

import QuiqChatClient from 'quiq-chat';

// Singleton chat client

let chatClient;

export const registerChatClient = (c: QuiqChatClient): void => {
  chatClient = c;
};

export const getChatClient = (): QuiqChatClient => {
  return chatClient;
};
