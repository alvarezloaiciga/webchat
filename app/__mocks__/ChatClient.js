export const registerChatClient = jest.fn();
export const getChatClient = jest.fn().mockReturnValue({
  onNewMessages: jest.fn().mockReturnThis(),
  onRegistration: jest.fn().mockReturnThis(),
  onAgentTyping: jest.fn().mockReturnThis(),
  onError: jest.fn().mockReturnThis(),
  onErrorResolved: jest.fn().mockReturnThis(),
  onConnectionStatusChange: jest.fn().mockReturnThis(),
  onNewSession: jest.fn().mockReturnThis(),
  onBurn: jest.fn().mockReturnThis(),
  onClientInactiveTimeout: jest.fn().mockReturnThis(),
  start: jest.fn(),
  stop: jest.fn(),
  emailTranscript: jest.fn(),
  getMessages: jest.fn(),
  joinChat: jest.fn(),
  leaveChat: jest.fn(),
  sendTextMessage: jest.fn(),
  updateMessagePreview: jest.fn(),
  sendRegistration: jest.fn(),
  checkForAgents: jest.fn(),
  hasTakenMeaningfulAction: jest.fn(),
  isRegistered: jest.fn().mockReturnValue(false),
  isChatVisible: jest.fn(),
  getChatConfiguration: jest.fn(),
  onSendTranscript: jest.fn().mockReturnValue({
    id: 1,
    type: 'SendTranscript',
    timestamp: 123456,
  }),
});
