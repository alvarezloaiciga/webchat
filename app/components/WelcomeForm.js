// @flow
import React from 'react';
import QUIQ from 'utils/quiq';
import HeaderMenu from 'HeaderMenu';
import {getDisplayString} from 'utils/i18n';
import {FormattedMessage} from 'react-intl';
import type {WelcomeFormField} from 'types';
import messages from 'messages';
import {sendRegistration} from 'quiq-chat';
import './styles/WelcomeForm.scss';

export type WelcomeFormProps = {
  onFormSubmit: (formattedString: string) => void,
  onPop: (fireEvent: boolean) => void,
  onDock: (fireEvent: boolean) => void,
  onMinimize: (fireEvent: boolean) => void,
};

const WelcomeForm = (props: WelcomeFormProps) => {
  const {WELCOME_FORM, FONT_FAMILY, COLOR} = QUIQ;

  // We shouldn't be rendering this component if we didn't find a WELCOME_FORM in the QUIQ object.
  // But just in case, pass it through so we don't block webchat.
  if (!WELCOME_FORM) {
    props.onFormSubmit('');
    return null;
  }

  const refs = {};
  const form = WELCOME_FORM;

  const renderField = (field: WelcomeFormField) =>
    <div className="field" key={field.id}>
      <label htmlFor={field.label} style={{fontFamily: FONT_FAMILY}}>
        {field.label}
        {field.required &&
          <span className="required" title={getDisplayString(messages.required)}> *</span>}
      </label>
      <input
        ref={n => (refs[field.id] = n)}
        type={field.type}
        name={field.id}
        required={field.required}
        style={{fontFamily: FONT_FAMILY}}
      />
    </div>;

  const submitForm = () => {
    const fields: {[string]: string} = {};
    Object.keys(refs).forEach(k => {
      fields[k] = refs[k].value;
    });
    sendRegistration(fields).then(props.onFormSubmit);
  };

  return (
    <div className="WelcomeForm">
      <HeaderMenu onPop={props.onPop} onDock={props.onDock} onMinimize={props.onMinimize} />
      <div className="welcomeFormBanner" style={{backgroundColor: COLOR}}>
        <span style={{fontFamily: FONT_FAMILY}}>{form.headerText}</span>
      </div>
      <div className="fields">
        {WELCOME_FORM.fields.map(renderField)}
      </div>
      <button
        className="submit"
        style={{background: COLOR, fontFamily: FONT_FAMILY}}
        onClick={submitForm}
      >
        <FormattedMessage {...messages.submitWelcomeForm} />
      </button>
    </div>
  );
};

export default WelcomeForm;
