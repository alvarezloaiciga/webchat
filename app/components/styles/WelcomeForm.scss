.WelcomeForm {
  height: 100%;
  display: flex;
  flex-direction: column;

  .welcomeFormBanner {
    text-align: center;
    flex: 0 0 auto;
    color: #fff;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    font-weight: 300;
    font-size: 14px;
    padding: 0 30px;
    background: #59ad5d;
    height: 80px;

    span {
      display: inline-block;
      width: 100%;
      line-height: 1.3;
    }
  }

  .formValidationError {
    margin: 20px;
    color: #c53431;
  }

  .fields {
    flex: 1 1 auto;
    display: flex;
    flex-direction: column;
    align-items: center;
    margin: 0 20px;
    overflow-y: auto;
  }

  .field,
  .submit {
    min-height: 30px;
  }

  .submit {
    flex: 0 0 auto;
    margin: 20px;
    width: 130px;
    align-self: flex-end;
    color: white;
    display: inline-block;
    vertical-align: middle;
    border: 1px solid transparent;
    font-size: 16px;
    text-transform: uppercase;
    line-height: 1;
    text-align: center;
    cursor: pointer;
    border-radius: 4px;
    padding: 8px;
    transition: all .15s ease-in-out;

    &:disabled {
      opacity: .7;
      cursor: not-allowed;
    }
  }

  .field {
    display: flex;
    flex-direction: column;
    align-self: stretch;
    align-items: stretch;
    min-width: 99%;
    flex: 0 0 auto;
    margin: 5px 0;

    label {
      margin-bottom: 4px;
    }

    label .required {
      color: #c53431;
      flex: 0 0 auto;
    }

    input {
      width: 100%;
      height: 40px;
      padding: 8px;
      box-sizing: border-box;
      border: 1px solid #cacaca;
      border-radius: 0;
      background-color: #fefefe;
      box-shadow: inset 0 1px 2px hsla(0, 0%, 4%, .1);
      font-family: inherit;
      font-size: 16px;
      font-weight: 400;
      color: #0a0a0a;
      transition: box-shadow .5s, border-color .25s ease-in-out;
    }
  }
}

// Total Height: 600px
// Banner: 100px
// Form: 450px
// Button 50px
.noFlexbox .WelcomeForm {
  .welcomeFormBanner {
    position: absolute;
    top: 0;
    height: 80px;
    min-height: 80px;
    max-height: 80px;
    width: 400px;
    min-width: 400px;
    max-width: 400px;
    display: table;
    color: white;
    padding: 5px;

    span {
      display: table-cell;
      vertical-align: middle;
    }
  }

  .fields {
    margin-top: 100px;
    min-height: 440px;
    max-height: 440px;
    height: 440px;

    .field {
      margin-left: 0;
      margin-right: 0;

      label {
        display: block;
        margin-bottom: 10px;
      }

      input {
        display: block;
        margin-bottom: 10px;
        width: 100%;
      }
    }

    .submit {
      margin-top: 0;
      float: right;
      clear: both;
    }
  }
}

.noFlexbox .WelcomeForm {
  .welcomeFormBanner {
    width: 392px;
    max-width: 392px;
    min-width: 392px;
    text-align: center;
    position: absolute;
    top: 30px;
    height: 50px;
    min-height: 50px;
    max-height: 50px;
    display: table;

    span {
      display: table-cell;
      vertical-align: middle;
    }
  }

  .submit {
    float: right;
  }
}
