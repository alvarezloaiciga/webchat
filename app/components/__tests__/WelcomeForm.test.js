// @flow
jest.mock('utils/quiq');
import React from 'react';
import type {WelcomeFormProps} from '../WelcomeForm';
import {WelcomeForm} from '../WelcomeForm';
import {shallow} from 'enzyme';
import type {ShallowWrapper} from 'enzyme';
import {registerChatClient} from '../../ChatClient';
import Textarea from 'react-textarea-autosize';

describe('WelcomeForm component', () => {
  let wrapper: ShallowWrapper;
  let testProps: WelcomeFormProps;
  let render: () => void;
  const mockClient = {
    sendMessage: jest.fn(() => {}),
    sendRegistration: jest.fn(async () => await {}),
  };

  beforeEach(() => {
    registerChatClient(mockClient);

    testProps = {};
    render = () => {
      wrapper = shallow(<WelcomeForm {...testProps} />);
    };
  });

  describe('rendering', () => {
    beforeEach(() => {
      render();
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    describe('rendering textareas', () => {
      it('renders textares correctly', () => {
        expect(wrapper.find(Textarea).length).toBe(1);
      });
    });

    describe('validation error rendering', () => {
      it('does not render error message if formValidationError state is false', () => {
        wrapper.setState({formValidationError: false});
        expect(wrapper.find('.formValidationError').length).toBe(0);
      });

      it('renders error message with formValidationError true', () => {
        wrapper.setState({formValidationError: true});
        expect(wrapper.find('.formValidationError').length).toBe(1);
      });
    });
  });

  describe('filling out form and submitting', () => {
    it('does not submit if there is a required field left blank', () => {
      render();
      wrapper.find('button').first().simulate('click', {preventDefault: jest.fn()});
      expect(mockClient.sendRegistration).not.toHaveBeenCalled();
    });

    it('does not submit if there is a required field containing only whitespace', () => {
      wrapper.find('input').at(0).simulate('change', {
        which: '   ',
        target: {
          name: 'firstName',
          value: '   ',
        },
      });

      wrapper.find('input').at(0).simulate('blur', {
        target: {
          name: 'firstName',
        },
      });
      render();
      wrapper.find('button').first().simulate('click', {preventDefault: jest.fn()});
      expect(mockClient.sendRegistration).not.toHaveBeenCalled();
    });

    describe('valid submission', () => {
      beforeEach(() => {
        wrapper.find('input').at(0).simulate('change', {
          which: 'a',
          target: {
            name: 'firstName',
            value: 'a',
          },
        });
        wrapper.find('button').first().simulate('click', {preventDefault: jest.fn()});
        render();
      });

      it('does submit if all required fields are filled in', () => {
        expect(mockClient.sendRegistration).toHaveBeenCalled();
      });

      it('does not send message if initial field was not provided', () => {
        expect(mockClient.sendMessage).not.toHaveBeenCalled();
      });

      it('disables send button and sets text', () => {
        expect(wrapper.find('button')).toMatchSnapshot();
      });
    });

    describe('valid submission with send message', () => {
      beforeEach(() => {
        wrapper.find('input').at(1).simulate('change', {
          which: 'brad',
          target: {
            name: 'lastName',
            value: 'brad',
          },
        });

        wrapper.find('input').at(0).simulate('change', {
          which: 'a',
          target: {
            name: 'firstName',
            value: 'a',
          },
        });

        wrapper.find('button').first().simulate('click', {preventDefault: jest.fn()});
        render();
      });

      it('does submit if all required fields are filled in', () => {
        expect(mockClient.sendRegistration).toHaveBeenCalled();
      });

      // Not sure why this is failing, but it is definitely being called...
      // it('does send message if initial field was provided', () => {
      //   expect(mockClient.sendMessage).toHaveBeenCalled();
      // });

      it('disables send button and sets text', () => {
        expect(wrapper.find('button')).toMatchSnapshot();
      });
    });
  });

  describe('form validation', () => {
    beforeEach(() => {
      render();
    });

    it('sets validationError state to true if a required field is left blank', () => {
      // $FlowIssue - flow think instance() returns a generic Component
      wrapper.instance().validateFormInput();
      expect(wrapper.state('formValidationError')).toBe(true);
    });
  });
});
