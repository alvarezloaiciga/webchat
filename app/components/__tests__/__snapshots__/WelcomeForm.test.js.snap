// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WelcomeForm component filling out form and submitting valid submission disables send button and sets text 1`] = `
<button
  className="submit"
  disabled={false}
  onClick={[Function]}
  style={
    Object {
      "background": "teal",
      "backgroundColor": "#333",
      "fontFamily": "Source Sans Pro",
      "fontSize": 16,
      "fontWeight": 700,
    }
  }
>
  Submit
</button>
`;

exports[`WelcomeForm component filling out form and submitting valid submission with send message disables send button and sets text 1`] = `
<button
  className="submit"
  disabled={false}
  onClick={[Function]}
  style={
    Object {
      "background": "teal",
      "backgroundColor": "#333",
      "fontFamily": "Source Sans Pro",
      "fontSize": 16,
      "fontWeight": 700,
    }
  }
>
  Submit
</button>
`;

exports[`WelcomeForm component rendering renders correctly 1`] = `
<form
  className="WelcomeForm"
  style={
    Object {
      "backgroundColor": "#f4f4f8",
    }
  }
>
  <Connect(HeaderMenu) />
  <div
    className="welcomeFormBanner"
    style={
      Object {
        "backgroundColor": "#333",
        "fontFamily": "Lato, sans-serif",
        "height": "auto",
        "padding": "8px 30px",
        "textAlign": "left",
      }
    }
  >
    Thanks for contacting us! Please fill out a couple brief pieces of information and we will get you chatting with an agent.
  </div>
  <Debugger />
  <div
    className="fields"
  >
    <div
      className="field"
      key="firstName"
      style={
        Object {
          "marginTop": 16,
        }
      }
    >
      <label
        htmlFor="First Name"
        style={
          Object {
            "color": "#555",
            "fontFamily": "Source Sans Pro",
            "fontWeight": 700,
            "textTransform": "uppercase",
          }
        }
      >
        First Name
        <span
          className="required"
          title="Required"
        >
           
          *
        </span>
      </label>
      <input
        maxLength={1000}
        name="firstName"
        onBlur={[Function]}
        onChange={[Function]}
        required={true}
        style={
          Object {
            "color": "#555",
            "fontFamily": "Source Sans Pro",
            "outline": "none",
          }
        }
        type="text"
        value=""
      />
    </div>
    <div
      className="field"
      key="lastName"
      style={
        Object {
          "marginTop": 16,
        }
      }
    >
      <label
        htmlFor="Last Name"
        style={
          Object {
            "color": "#555",
            "fontFamily": "Source Sans Pro",
            "fontWeight": 700,
            "textTransform": "uppercase",
          }
        }
      >
        Last Name
      </label>
      <input
        maxLength={1000}
        name="lastName"
        onBlur={[Function]}
        onChange={[Function]}
        required={false}
        style={
          Object {
            "color": "#555",
            "fontFamily": "Source Sans Pro",
            "outline": "none",
          }
        }
        type="text"
        value=""
      />
    </div>
    <div
      className="field"
      key="numberField"
      style={
        Object {
          "marginTop": 16,
        }
      }
    >
      <label
        htmlFor="Number Field"
        style={
          Object {
            "color": "#555",
            "fontFamily": "Source Sans Pro",
            "fontWeight": 700,
            "textTransform": "uppercase",
          }
        }
      >
        Number Field
      </label>
      <input
        maxLength={1000}
        name="numberField"
        onBlur={[Function]}
        onChange={[Function]}
        required={false}
        style={
          Object {
            "color": "#555",
            "fontFamily": "Source Sans Pro",
            "outline": "none",
          }
        }
        type="number"
        value=""
      />
    </div>
    <div
      className="field"
      key="email"
      style={
        Object {
          "marginTop": 16,
        }
      }
    >
      <label
        htmlFor="E-Mail"
        style={
          Object {
            "color": "#555",
            "fontFamily": "Source Sans Pro",
            "fontWeight": 700,
            "textTransform": "uppercase",
          }
        }
      >
        E-Mail
      </label>
      <input
        maxLength={1000}
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        required={false}
        style={
          Object {
            "color": "#555",
            "fontFamily": "Source Sans Pro",
            "outline": "none",
          }
        }
        type="email"
        value=""
      />
    </div>
    <div
      className="field"
      key="phoneNumber"
      style={
        Object {
          "marginTop": 16,
        }
      }
    >
      <label
        htmlFor="Phone Number"
        style={
          Object {
            "color": "#555",
            "fontFamily": "Source Sans Pro",
            "fontWeight": 700,
            "textTransform": "uppercase",
          }
        }
      >
        Phone Number
      </label>
      <input
        maxLength={1000}
        name="phoneNumber"
        onBlur={[Function]}
        onChange={[Function]}
        required={false}
        style={
          Object {
            "color": "#555",
            "fontFamily": "Source Sans Pro",
            "outline": "none",
          }
        }
        type="tel"
        value=""
      />
    </div>
    <div
      className="field"
      key="lifeStory"
      style={
        Object {
          "marginTop": 16,
        }
      }
    >
      <label
        htmlFor="My life story"
        style={
          Object {
            "color": "#555",
            "fontFamily": "Source Sans Pro",
            "fontWeight": 700,
            "textTransform": "uppercase",
          }
        }
      >
        My life story
      </label>
      <textarea
        maxLength={1000}
        name="lifeStory"
        onBlur={[Function]}
        onChange={[Function]}
        required={false}
        resize="vertical"
        rows={3}
        style={
          Object {
            "color": "#888",
            "fontFamily": "Source Sans Pro",
            "outline": "none",
          }
        }
        value=""
      />
    </div>
    <div
      className="field"
      key="country"
      style={
        Object {
          "marginTop": 16,
        }
      }
    >
      <label
        htmlFor="Country"
        style={
          Object {
            "color": "#555",
            "fontFamily": "Source Sans Pro",
            "fontWeight": 700,
            "textTransform": "uppercase",
          }
        }
      >
        Country
        <span
          className="required"
          title="Required"
        >
           
          *
        </span>
      </label>
      <select
        name="country"
        onBlur={[Function]}
        onChange={[Function]}
        required={true}
        style={
          Object {
            "color": "#888",
            "fontFamily": "Source Sans Pro",
            "outline": "none",
          }
        }
        value=""
      >
        <option
          key="---"
          value=""
        >
          ---
        </option>
        <option
          key="United StatesUSA"
          value="USA"
        >
          United States
        </option>
        <option
          key="CanadaCanada"
          value="Canada"
        >
          Canada
        </option>
      </select>
    </div>
  </div>
  <button
    className="submit"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "background": "teal",
        "backgroundColor": "#333",
        "fontFamily": "Source Sans Pro",
        "fontSize": 16,
        "fontWeight": 700,
      }
    }
  >
    Submit
  </button>
</form>
`;
